datashader/core.py

```diff
@@ -196,11 +196,121 @@ def line(self, source, x, y, agg=None):
+
+        cols = vertices.columns.tolist()
+        x, y, weights = cols[0], cols[1], cols[2:]
+
+        # Verify the simplex data structure
+        assert simplices.values.shape[1] >= 3, ('At least three vertex columns '
+                                                'are required for the triangle '
+                                                'definition')
+        simplices_all_ints = simplices.dtypes.iloc[:3].map(
+            lambda dt: np.issubdtype(dt, np.integer)
+        ).all()
+        assert simplices_all_ints, ('Simplices must be integral. You may '
+                                    'consider casting simplices to integers '
+                                    'with ".astype(int)"')
+
+        # Choose reduction based on vertex coordinate data
+        weight_col = None
+        verts_have_weights = not not weights
+        if verts_have_weights:
+            weight_col = weights[0]
+            if agg is None:
+                agg = mean_rdn(weight_col)
+            elif agg.column is None:
+                # Rasterization code assumes presence of agg column
+                agg.column = weight_col
+        else:
+            assert simplices.values.shape[1] > 3, 'If no vertex weight column is provided, a triangle weight column is required.'
+            weight_col = simplices.columns[3]
+            if agg is None:
+                agg = mean_rdn(weight_col)
+            elif agg.column is None:
+                # Rasterization code assumes presence of agg column
+                agg.column = weight_col
+
+        if mesh is None:
+            if (isinstance(vertices, dd.DataFrame) and
+                    isinstance(simplices, dd.DataFrame)):
+                # TODO: For dask: avoid .compute() calls, and add winding auto-detection
+                vertex_idxs = simplices.values[:, :3].astype(np.int64)
+                vals = vertices.values.compute()[vertex_idxs]
+                vals = vals.reshape(np.prod(vals.shape[:2]), vals.shape[2])
+                source = pd.DataFrame(vals, columns=vertices.columns)
+                if not verts_have_weights:
+                    source[weight_col] = simplices.values[:, 3].compute().repeat(3)
+            else:
+                winding = [0, 1, 2]
+                # Change vertex winding to CW if necessary
+                first_tri = vertices.values[simplices.values[0, winding].astype(np.int64), :2]
+                a, b, c = first_tri
+                if np.cross(b-a, c-a).item() >= 0:
+                    winding = [0, 2, 1]
+
+                vertex_idxs = simplices.values[:, winding].astype(np.int64)
+                vals = vertices.values[vertex_idxs]
+                vals = vals.reshape(np.prod(vals.shape[:2]), vals.shape[2])
+                source = pd.DataFrame(vals, columns=vertices.columns)
+                if not verts_have_weights:
+                    source[weight_col] = simplices.values[:, 3].repeat(3)
+        else:
+            source = mesh
```

Comment:
How is the mesh argument meant to be used?  The docstring doesn't say enough to make it usable, and I would have thought that the code here would all be in some external function that would be called here but which the docstring could suggest that the user call explicitly if they want to avoid having to redo this step.

Reply:
+1 on providing a function that computes ``mesh`` for you. Doesn't have to be in this PR though.

Reply:
Thanks for the feedback. I'll revise the docstring, and provide a utility function to create the mesh data structure, in my next PR.